tools/topology/sof-smart-amplifier.m4

```diff
@@ -225,7 +225,7 @@ DAI_CONFIG(SSP, SMART_SSP_INDEX, SMART_BE_ID, SMART_SSP_NAME,
 		      SSP_CLOCK(bclk, 12288000, codec_slave),
 		      SSP_CLOCK(fsync, 48000, codec_slave),
 		      SSP_TDM(8, 32, 15, 255),
-		      SSP_CONFIG_DATA(SSP, SMART_SSP_INDEX, 24, 0, SMART_SSP_QUIRK)))
```

Comment:
@sathya-nujella is there any explanation for this fix? the TDM slot is already 32 bits, why does this matter if we put 32-bit or 24-bit data in the slot? Something's not adding up. If there is noise added, this points to a distortion inside of the firmware.

Not buying this fix, or there's something else not explained.

Reply:
BCLK is configured for 32-bit (32*8*48=12.288MHz).

In DAI config, if it is set to S24_LE then during playback:
call stack which will get called during playback which sets the correct bit depth.
[   59.720087]  max98373_dai_hw_params+0x67/0x3bb [snd_soc_max98373_i2c]
[   59.720091]  sof_pcm_dai_link_fixup+0x34/0x204 [snd_sof]
[   59.720093]  snd_soc_dai_hw_params+0x4f/0x91
[   59.720097]  soc_pcm_hw_params+0x2db/0x586
[   59.720104]  dpcm_be_dai_hw_params+0x197/0x29f
[   59.720107]  dpcm_fe_dai_hw_params+0x9e/0x18d

because of this, Maxim registers are set to 24 bit. So, there is a mis-match in BCLK and maxim Bit-depth register configuration. 
PIPELINE_PCM_ADD supports up to S32-LE, so issue is not with FE. But with BCLK and DAI configuration (DAI_ADD entries) in topology.

Reply:
Are you saying the dai s24le/s32le configurations line 171/178 is propagated to the codec drivers?

Reply:
yes, during playback, I dumped call stack and see this is getting propagated to: max98373_dai_ops -> hw_params

https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/codecs/max98373-i2c.c#L400

Reply:
ack, thanks @sathya-nujella for the explanations - and educating me on what parameters are propagated to codec drivers.